virtualmachine/gcp/vm.go

```diff
@@ -206,10 +335,366 @@ func (vm *VM) DeleteDisks() error {
+	if err != nil {
+		return nil, err
+	}
+
+	response := make([]Zone, 0)
+
+	for _, zone := range zoneList {
+		regionName := convResURLToName(zone.Region)
+		response = append(response, Zone{
+			Name:        zone.Name,
+			Description: zone.Description,
+			Region:      regionName,
+			Status:      zone.Status,
+		})
+	}
+
+	return response, nil
+}
+
+// GetRegionList gets the list of available regions
+func (vm *VM) GetRegionList() ([]Region, error) {
+	s, err := vm.getService()
+	if err != nil {
+		return nil, err
+	}
+
+	regionList, err := s.getRegionList()
+	if err != nil {
+		return nil, err
+	}
+
+	response := make([]Region, 0)
+
+	zones := make([]string, 0)
+	for _, region := range regionList {
+		zones = nil
+		for _, zoneUrl := range region.Zones {
+			zones = append(zones, convResURLToName(zoneUrl))
+		}
+		response = append(response, Region{
+			Name:        region.Name,
+			Description: region.Description,
+			Zones:       zones,
+			Status:      region.Status,
+		})
+	}
+
+	return response, nil
+}
+
+// AddNewDisks adds a new disks to an instance
+func (vm *VM) AddNewDisks() ([]AttachedDisk, error) {
+	disksPresent := make(map[string]bool)
+	response := make([]AttachedDisk, 0)
+
+	s, err := vm.getService()
+	if err != nil {
+		return nil, err
+	}
+
+	for _, disk := range vm.Disks {
+		if err := s.createDisk(&disk); err != nil {
+			return nil, err
+		}
+
+		if err := s.attachDisk(&disk); err != nil {
+			return nil, err
+		}
+		disksPresent[disk.Name] = true
+	}
+
+	instance, err := s.getInstance()
+	if err != nil {
+		return nil, err
+	}
+
+	for _, attachedDisk := range instance.Disks {
+		if disksPresent[attachedDisk.DeviceName] {
+			response = append(response, AttachedDisk{
+				Name:       attachedDisk.DeviceName,
+				Boot:       &attachedDisk.Boot,
+				AutoDelete: &attachedDisk.AutoDelete,
+				Interface:  attachedDisk.Interface,
+				DevicePath: DevicePathPrefix +
+					attachedDisk.DeviceName,
+			})
+		}
+	}
+	return response, nil
+}
+
+// IsInstance checks if the given instance is present in GCP. Returns true if
+// instance is available else false.
+func (vm *VM) IsInstance() (bool, error) {
+	s, err := vm.getService()
+	if err != nil {
+		return false, err
+	}
+
+	_, err = s.getInstance()
+	if err != nil {
+		return false, err
+	} else {
+		return true, nil
+	}
+}
+
+// DeleteVMDisk detaches the given disks from the instance and deletes them
+func (vm *VM) DeleteVMDisks() error {
+	s, err := vm.getService()
+	if err != nil {
+		return err
+	}
+
+	for _, disk := range vm.Disks {
+		if err := s.detachDisk(&disk); err != nil {
+			return err
+		}
+
+		if err := s.deleteDisk(disk.Name); err != nil {
+			return err
+		}
+	}
+	return nil
+}
+
+// GetInstance gets instance details
+func (vm *VM) GetInstance() (*InstanceData, error) {
+	s, err := vm.getService()
+	if err != nil {
+		return nil, err
+	}
+
+	instance, err := s.getInstance()
+	if err != nil {
+		return nil, err
+	}
+	ips, err := vm.GetIPs()
```

Comment:
Will it also wait for IP allocation or access? with timeout?

Reply:
GetIPs() does not wait. However, for every action we take we are waiting for that action to be complete with waitForOperation() in util.go. waitForOperation will return only if the action such as provision is complete or fails. If the action is complete, then provision would have allocated an IP to it.